src/backend/distributed/master/master_metadata_utility.c

```diff
@@ -62,6 +66,249 @@ static ShardPlacement * TupleToShardPlacement(TupleDesc tupleDesc,
 /* exports for SQL callable functions */
 PG_FUNCTION_INFO_V1(master_stage_shard_row);
 PG_FUNCTION_INFO_V1(master_stage_shard_placement_row);
+PG_FUNCTION_INFO_V1(citus_table_size);
+PG_FUNCTION_INFO_V1(citus_total_relation_size);
+PG_FUNCTION_INFO_V1(citus_relation_size);
+
+/*
+ * GetShardIntervalsOnNode takes WorkerNode then compare it with each placement of table.
+ * It returns shard intervals of table on that node as a list of shard intervals.
+ * Note that, shard intervals returned as elements of the list are not the copies but the
+ * pointers. DO NOT modify the shard intervals returned by this function.
+ */
+List *
+GetShardIntervalsOnWorkerNode(WorkerNode *workerNode, Oid relationId)
+{
+	DistTableCacheEntry *distTableCacheEntry = DistributedTableCacheEntry(relationId);
+	char *workerNodeName = workerNode->workerName;
+	uint32 workerNodePort = workerNode->workerPort;
+	List *shardIntervalList = NIL;
+	int shardIndex = 0;
+	int shardIntervalArrayLength = distTableCacheEntry->shardIntervalArrayLength;
+
+	for (shardIndex = 0; shardIndex < shardIntervalArrayLength; shardIndex++)
+	{
+		ShardPlacement *placementArray =
+			distTableCacheEntry->arrayOfPlacementArrays[shardIndex];
+		int numberOfPlacements =
+			distTableCacheEntry->arrayOfPlacementArrayLengths[shardIndex];
+		int placementIndex = 0;
+
+		for (placementIndex = 0; placementIndex < numberOfPlacements; placementIndex++)
+		{
+			ShardPlacement *placement = &placementArray[placementIndex];
+			char *shardNodeName = placement->nodeName;
+			uint32 shardNodePort = placement->nodePort;
+			uint64 shardId = placement->shardId;
+
+			LockShardDistributionMetadata(shardId, AccessShareLock);
```

Comment:
We would need to obtain this lock prior to calling DistributedTableCacheEntry. The lock being held means a transaction that is changing the metadata is in progress. Once that transaction is done and the lock is released, we should re-read the metadata. 

We could also consider skipping the lock, since we don't really want to block `citus_table_size` for very long. The worst that could happen is that this function errors out when the placement is dropped after a move. Alternatively, we could use `TryLockShardDistributionMetadata` and skip over the placement if it is being moved and print a warning that it is being excluded from the result.